src/checkout.c

```diff
@@ -2397,20 +2397,20 @@ static int checkout_data_init(
 				GIT_DIR_MODE, GIT_MKDIR_VERIFY_DIR)) < 0)
 		goto cleanup;
 
+	if ((error = git_repository_index(&data->index, data->repo)) < 0)
+		goto cleanup;
```

Comment:
So, previously we were not loading the index into `data->index` when `GIT_CHECKOUT_NO_REFRESH` was set. This leaves me wondering about line 2441,  where we unconditionally dereference `data->index`. Was this code broken previously and would segfault in case `GIT_CHECKOUT_NO_REFRESH` was set or is the index already loaded at this point in time, resulting in a memory leak due to an additional reference to the index?

Reply:
Yes, it would have dereferenced that `NULL` index.  the `NO_REFRESH` bit is really not a great pattern to use, so I'm not surprised that nobody's ever noticed this.  However, I added a test to make sure that we don't regress again.